### ISSUE TITLE
expected behavior of TCP keepalive

### ISSUE THREAD

**Issue by alberanid** [baresip/re#288]
Hi,
I have some systems where, apparently randomly (after many hours, usually) and without other network failures, the connection is lost and is re-established only when a re-register happens after the configured regint time. This leaves some minutes when the the system is not reachable.

The account:
```
"user" <sip:user@sip.example.net;transport=tcp>;auth_user=user;auth_pass=password;answermode=manual;audio_codecs=PCMA,PCMU;regint=600;pubint=300;video_codecs=h264;outbound="sip:sip.example.net;transport=tcp";sipnat=outbound
```

As you can see I'm using sipnat outbound.

In this case, looking at the code, I assument it should enable a TCP keepalive that if I've understood correctly, will publish a `\r\n` every 120 seconds (TCP_KEEPALIVE_INTVAL) and wait 10 seconds before closing a non-working connection: https://github.com/baresip/re/blob/c7599c186b0edbdd0032ac32914bdbf852cee26f/src/sip/transp.c#L1829

However, from a tcpdump I do not see any '\r\n' lines sent at any interval.

Should I look for something else? Is this a feature that requires server-side support? (we are currently using FlexiSIP)

Thanks for your help!

**Comment by sreimers** [comment]
Can you try to add a debug msg to the keepalive handler:

```diff
--- a/src/sip/transp.c
+++ b/src/sip/transp.c
@@ -316,6 +316,8 @@ static void conn_keepalive_handler(void *arg)
        mb.pos  = 0;
        mb.end  = 4;
 
+       DEBUG_WARNING("KEEPALIVE\n");
+
        err = tcp_send(conn->tc, &mb);
        if (err) {
                conn_close(conn, err);
```

You can do the same for `sip_keepalive_tcp` to see if the timer handler setup is working.

**Comment by alfredh** [comment]
in general it is probably better to use the mailing list for questions like this.

I think the sip outbound ping/pong feature is only enabled if supported by the SIP Server.
Please check in the source code.

You can also use a low registration interval to keep the TCP connection open:

```
;regint=60
```

please note that some networks (CGN) will kill random TCP connections at random times.
If you want this to work reliably you need to re-connection TCP from the client side.

**Comment by alberanid** [comment]
Agreed, thanks for the support; I'll let you know if we will detect anything meaningful.
